File: JavaScript/ExampleTests/common/AdderString.js
import * as tokens from "./tokens.js";

export class AdderString {
  static DNC_STRING = "?DNC?";

  constructor(number1 = "", number2 = "", result = "") {
    this.number1 = number1;
    this.number2 = number2;
    this.result = result;
  }

  static fromList(values) {
    const v = Array.from(values);
    return new AdderString(
      v[0] ?? "",
      v[1] ?? "",
      v[2] ?? ""
    );
  }

  /** Builds from the text form, e.g. Money as "25 USD". */
  static fromText(text) {
    const parts = tokens.require_(text, 3, "Adder");
    return new AdderString(
      parts[0],
      parts[1],
      parts[2]
    );
  }

  toString() {
    return tokens.token(this.number1) + " " + tokens.token(this.number2) + " " + tokens.token(this.result);
  }

  equals(other) {
    if (!(other instanceof AdderString)) return false;
    return (this.number1 === AdderString.DNC_STRING || other.number1 === AdderString.DNC_STRING || this.number1 === other.number1)
      && (this.number2 === AdderString.DNC_STRING || other.number2 === AdderString.DNC_STRING || this.number2 === other.number2)
      && (this.result === AdderString.DNC_STRING || other.result === AdderString.DNC_STRING || this.result === other.result);
  }

  static equalLists(a, b) {
    if (a.length !== b.length) return false;
    return a.every((row, i) => row.equals(b[i]));
  }
}

File: Python/ExampleTests/common/adder_string.py
from . import tokens
DNC_STRING = '?DNC?'


class AdderString:
    def __init__(self, number1: str = '', number2: str = '', result: str = ''):
        self.number1 = number1
        self.number2 = number2
        self.result = result

    @classmethod
    def from_list(cls, values):
        v = list(values)
        return cls(
            v[0] if len(v) > 0 else '',
            v[1] if len(v) > 1 else '',
            v[2] if len(v) > 2 else ''
        )

    @classmethod
    def from_text(cls, text):
        """Builds from the text form, e.g. Money as '25 USD'."""
        parts = tokens.require(text, 3, 'Adder')
        return cls(
            parts[0],
            parts[1],
            parts[2]
        )

    def __str__(self):
        return (tokens.token(self.number1) + ' ' +
                tokens.token(self.number2) + ' ' +
                tokens.token(self.result))

    def _key(self):
        return (self.number1, self.number2, self.result)

    def __eq__(self, other):
        if not isinstance(other, AdderString):
            return NotImplemented
        return all(a == b or a == DNC_STRING or b == DNC_STRING
                   for a, b in zip(self._key(), other._key()))

    def __hash__(self):
        return hash(self._key())

    def __repr__(self):
        return f'AdderString({self})'
